File: openacc-wg-2016/matmul_acc.c
#include <stdio.h>
#include "ticktock.h"

#define SIZE 2250

int a[SIZE][SIZE];
int b[SIZE][SIZE];
int c[SIZE][SIZE];

void matmul(){
    int i,j,k;
    #pragma acc kernels
    for(i=0;i<SIZE;i++)
        for(j=0;j<SIZE;j++){
            int sum = 0;
            for(k=0;k<SIZE;k++)
                sum += a[i][k] * b[k][j];
            c[i][j] = sum;
        }
}

int main(){
    int i,j;

    for(i=0;i<SIZE;i++)
        for(j=0;j<SIZE;j++){
            a[i][j] = (i % 50);
            b[i][j] = (j % 30);
            c[i][j] = 0;
        }
    tick_tock tt;
    tick(&tt); 
    matmul();
    tock(&tt);    
    printf("c[523][789] = %d\n",c[523][789]);
    return 0;     
}

File: openacc-wg-2016/matmul_acc.f90
program matmul_acc
    
    implicit none
    integer,parameter :: SIZE=1000

    real :: a(SIZE,SIZE), b(SIZE,SIZE), c(SIZE,SIZE);
    real :: sum
    integer :: i,j,k
    integer :: it0,it1

    do i=1,SIZE
        do j=1,SIZE
            a(i,j) = real(i+j);
            b(i,j) = real(i-j);
            c(i,j) = 0.0;
        end do
    end do
    call tick(it0,it1) 
    !$acc kernels 
    do i=1,SIZE
        do j=1,SIZE
            sum = 0.0
            do k=1,SIZE
                sum = sum + a(i,k) * b(k,j)
            end do
            c(i,j) = sum
        end do
    end do
    !$acc end kernels
    call tock(it0,it1)    
end program matmul_acc
